### PR TITLE
MQE-1800: Allow generate:tests to continue running even if one test f…

### PR DESCRIPTION
**Pull request by jilu1**
…ails generation

### Fixed Issues (if relevant)

1. magento/magento2-functional-testing-framework#<issue_number>: Issue title
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

### CODE REVIEW

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php [hunk 1]
```diff
@@ -94,27 +95,34 @@ public function generateAllSuites($testManifest)
     {
         $suites = $testManifest->getSuiteConfig();
 
+        $exceptionCollector = new ExceptionCollector();
         foreach ($suites as $suiteName => $suiteContent) {
-            if (empty($suiteContent)) {
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:98]: I see these notices when I run generate:tests command. Are they expected?

NOTICE: Suite 'WYSIWYGDisabledSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoProductionModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoDeveloperModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'SearchEngineElasticsearchSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'InContextPaypalSuite' contains no tests and won't be generated.
[]

File: src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php [hunk 2]
```diff
@@ -72,6 +72,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
         }
 
-        $output->writeLn("Suites Generated");
+        if ($this->cmdStatus) {
+            $output->writeLn("Suites Generated");
+            return 0;
+        } else {
+            $output->writeLn("Suite parsing error found. See mftf.log for details.");
+            $output->writeLn("Suites Generated");
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php:80]: to indicate all tests in suite didn't generate, can the error message be tweaked:
`Suites Generated (partial)` or
`Suites Generated  (with failures)`

File: src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php [hunk 3]
```diff
@@ -152,7 +161,16 @@ protected function execute(InputInterface $input, OutputInterface $output)
             return 1;
         }
 
-        $output->writeln("Generate Tests Command Run");
+        if (!empty($errMessages)) {
+            foreach (array_unique($errMessages) as $errMessage) {
+                $output->writeln($errMessage);
+                $output->writeln("\nGenerate Tests Command Run");
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php:167]: to indicate all tests didn't generate, can the error message be tweaked:
`Generate Tests Command Run (partial)` or
`Generate Tests Command Run (with failures)`

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php [hunk 4]
```diff
@@ -63,36 +63,75 @@ public function parseSuiteDataIntoObjects($parsedSuiteData)
...
-            $groupTestsToExclude = $parsedSuite[self::EXCLUDE_TAG_NAME] ?? [];
+                // resolve references as test objects
+                // continue if failed in include
+                $include = $this->extractTestObjectsFromSuiteRef($groupTestsToInclude);
+                $includeTests = $include['objects'] ?? [];
+                $stepError = $include['status'] ?? 0;
+                $includeMessage = '';
+                if ($stepError != 0) {
+                    $noError = false;
+                    $includeMessage = strval($stepError) . " test(s) not included for suite \""
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:89]: I see this error when running generate:tests though i see the test generated. 
`1 test(s) not included for suite "AppConfigDumpSuite"`

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php [hunk 5]
```diff
@@ -239,10 +255,17 @@ public function createAllTestFiles($testManifest = null, $testsToIgnore = null)
...
 
-        $testPhpArray = $this->assembleAllTestPhp($testManifest, $testsToIgnore);
-        foreach ($testPhpArray as $testPhpFile) {
+        $noError = $this->assembleAllTestPhp($testManifest, $testsToIgnore);
+        foreach ($this->cestPhpArray as $testPhpFile) {
             $this->createCestFile($testPhpFile[1], $testPhpFile[0]);
         }
+
+        if (!$noError) {
+            throw new TestFrameworkException("ERROR: " .
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:264]: Can this error message be unified across all error types:
- schema validation error
- annotation errors

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php [hunk 6]
```diff
@@ -319,45 +342,67 @@ private function generateInjectMethod()
...
 
-            //write to manifest here if manifest is not null
-            if ($testManifest != null) {
-                $testManifest->addTest($test);
+                //write to manifest here if manifest is not null
+                if ($testManifest != null) {
+                    $testManifest->addTest($test);
+                }
+            } catch (\Exception $e) {
+                $this->notGeneratedTestsArray[] = [$test->getName() => $e->getMessage()];
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:391]: can number of tests that failed to generate due to schema validation error and annotation errors be printed in a similar format in mftf log?

**Review by soumyau — CHANGES_REQUESTED** [review]

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 4 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:88]: Can you check again? Please add command used to see the error.

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 1 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:98]: Can you check again? Please add command used to see the error.

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 6 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:402]: It's print in similar way with `number + word`. Let me know if you mean something else.

File: src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php (hunk 3 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php]: Liked the 2nd version. Fixed

File: src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php (hunk 2 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php]: Liked the 2nd version. Changed

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 5 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php]: Schema errors will still break tests.
Annotation errors will not break tests as before.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 4 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:88]: 1. I introduced a schema error in `AdminLoginSuccessfulTest` for this step -
`<actionGroup1 ref="AssertAdminSuccessLoginActionGroup" stepKey="assertLoggedIn"/>`

2. then I ran `bin/mftf generate:tests`

this is the output I get:
```
FAILURE: Schema validation error Element 'actionGroup1': This element is not expected.Line: 25
Unable to parse suite "SecureStorefrontURLSuite"NOTICE: 164 Section name violations detected. See mftf.log for details.[]
NOTICE: 617 section element name violations detected. See mftf.log for details.[]
NOTICE: 301 data entity name violations detected. See mftf.log for details.[]
NOTICE: 3 data entity key violations detected. See mftf.log for details.[]
1 test(s) not included for suite "AppConfigDumpSuite"NOTICE: 10 action group name violations detected. See mftf.log for details.[]
NOTICE: 35 Page name violations detected. See mftf.log for details.[]
ERROR: 2 Suite failed to generate. See mftf.log for details.NOTICE: Suite 'AppConfigDumpSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'WYSIWYGDisabledSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoProductionModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoDeveloperModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'SearchEngineElasticsearchSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'InContextPaypalSuite' contains no tests and won't be generated.
```

Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:88]: 1. I introduced a schema error in `AdminLoginSuccessfulTest` for this step -
`<actionGroup1 ref="AssertAdminSuccessLoginActionGroup" stepKey="assertLoggedIn"/>`

2. then I ran `bin/mftf generate:tests`

this is the output I get:
```
FAILURE: Schema validation error Element 'actionGroup1': This element is not expected.Line: 25
Unable to parse suite "SecureStorefrontURLSuite"NOTICE: 164 Section name violations detected. See mftf.log for details.[]
NOTICE: 617 section element name violations detected. See mftf.log for details.[]
NOTICE: 301 data entity name violations detected. See mftf.log for details.[]
NOTICE: 3 data entity key violations detected. See mftf.log for details.[]
1 test(s) not included for suite "AppConfigDumpSuite"NOTICE: 10 action group name violations detected. See mftf.log for details.[]
NOTICE: 35 Page name violations detected. See mftf.log for details.[]
ERROR: 2 Suite failed to generate. See mftf.log for details.NOTICE: Suite 'AppConfigDumpSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'WYSIWYGDisabledSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoProductionModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'MagentoDeveloperModeOnlyTestSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'SearchEngineElasticsearchSuite' contains no tests and won't be generated.
[]
NOTICE: Suite 'InContextPaypalSuite' contains no tests and won't be generated.
```

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 1 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:98]: I see this when I introduce a schema error in `AdminLoginSuccessfulTest`, to see if everything else gets generated except for this test. Also noticed that schema validation error causes test generation to fail, though I see this message `Generate Tests Command Run`. The generated folder is empty.

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 6 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:402]: I introduced a schema error in `AdminLoginFailedTest` 

I was expecting to also see something like the below:
1. Error: Failed to generate {$test->getName()}
2. All tests generated except for AdminLoginFailedTest


What I see:
1. FAILURE: Schema validation error Element 'actionGroup1': This element is not expected.Line: 23
2. No tests generated

Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:402]: I introduced an annotation error by copying annotation block from `AdminLoginSuccessfulTest` to `AdminLoginFailedTest` (duplicate error)
I was expecting to see something like the below:
1. 2 tests failed to generate.
2. All tests generated except for `AdminLoginFailedTest` and `AdminLoginSuccessfulTest`

What I see:
1. ERROR: 1 Suite failed to generate. See mftf.log for details.
2. No tests generated

**Review by soumyau — CHANGES_REQUESTED** [review]

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 5 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php]: Does that mean:

- schema error results in no test generated? 
- annotation error cause the tests to still generate with erroneous annotations? 

Also from AC:
> Schema errors do not cause hard failure
> Annotation errors do not cause hard failure

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 6 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:402]: What I see:
All tests generate when there is an annotation error
No tests generate when there is a schema validation error

I was expecting:

- All tests except the test xmls with annotation error generate (will need to evaluate if test runs with incorrect annotations can cause any run time failures)
- All tests except the test xmls with schema validation error generate (if this can't be achieved, maybe a hard failure is still the better solution)

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 1 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:98]: Fixed

**Comment by jilu1** [comment]
Multiple comments around the same issue, I will summarize it here:

1) It's impossible to continue when schema validation error happens. We don't even know what file it is at the point. We must have a hard fail. It was broken by my earlier change and it should have been fixed now.
2) There is no change for annotation errors. If they are not failing generation previously and should not fail now in theory. Let me know if you find anything differently. We are not changing the behavior because there are hundreds of annotations errors and there is no way to generate tests without them.

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 6 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:406]: see reply in general comment

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 5 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php]: see reply in general comment

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 4 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:94]: Fixed

**Comment by jilu1** [comment]
UR Build:
https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/35825/

**Review by soumyau — CHANGES_REQUESTED** [review]
Added observations and issues found and some rephrasing of error messages:

It would be good to add additional verification test coverage for newly added/modified error messages wherever applicable.

Add jenkins build results for failure conditions (make some modifications to existing tests by introducing errors in tests and suites and verify console and logs) and also for pass conditions.

It will be helpful to have some documentation for Generate Tests Run Command (with errors) here - https://devdocs.magento.com/mftf/docs/commands/mftf.html (or any appropriate page)
Giving a high level overview of:
1. which type of errors will result in hard failures
2. which errors will be written into mftf.log but will still generate unaffected tests

File: src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php [hunk 7]
```diff
@@ -66,12 +68,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
         }
 
-        $output->writeLn("Suites Generated");
+        if (empty(GenerationErrorHandler::getInstance()->getAllErrors())) {
+            $output->writeln("Suites Generated");
+            return 0;
+        } else {
+            GenerationErrorHandler::getInstance()->printErrorSummary();
+            GenerationErrorHandler::getInstance()->reset();
+            $output->writeln("Suites Generated (with failures)");
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php:88]: Changed

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php [hunk 8]
```diff
@@ -71,28 +81,80 @@ public function parseSuiteDataIntoObjects($parsedSuiteData)
...
-            $includeTests = $this->extractTestObjectsFromSuiteRef($groupTestsToInclude);
-            $excludeTests = $this->extractTestObjectsFromSuiteRef($groupTestsToExclude);
+                // resolve references as test objects
+                // continue if failed in include
+                $include = $this->extractTestObjectsFromSuiteRef($groupTestsToInclude);
+                $includeTests = $include['objects'] ?? [];
+                $stepError = $include['status'] ?? 0;
+                $includeMessage = '';
+                if ($stepError != 0) {
+                    $includeMessage = "ERROR: " . strval($stepError) . " test(s) not included for suite "
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:96]: > maybe reword to:
> test(s) not defined in xml for suite

This error is for tests defined in xml but has error and end up not being included.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php [hunk 9]
```diff
@@ -144,24 +206,32 @@ private function validateSuiteName($parsedSuite)
...
 
         if (array_key_exists(TestObjectExtractor::TEST_BEFORE_HOOK, $parsedSuite)) {
-            $suiteHooks[TestObjectExtractor::TEST_BEFORE_HOOK] = $this->testHookObjectExtractor->extractHook(
+            $hookObject = $this->testHookObjectExtractor->extractHook(
                 $parsedSuite[self::NAME],
                 TestObjectExtractor::TEST_BEFORE_HOOK,
                 $parsedSuite[TestObjectExtractor::TEST_BEFORE_HOOK]
             );
+            // Validate hook actions
+            $hookObject->getActions();
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:222]: This is to validate the correctness of the hook so that it won't break the entire generation.

File: src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php [hunk 10]
```diff
@@ -92,12 +97,32 @@ public function getObject($testName)
...
+            try {
+                $testObjects[$testName] = $this->extendTest($test);
+            } catch (FastFailException $exception) {
+                throw $exception;
+            } catch (\Exception $exception) {
+                LoggingUtil::getInstance()->getLogger(self::class)->error(
+                    "Unable to create test " . $testName . "\n" . $exception->getMessage()
+                );
+                if (MftfApplicationConfig::getConfig()->getPhase() == MftfApplicationConfig::GENERATION_PHASE) {
+                    print("ERROR: Unable to create test " . $testName . "\n" . $exception->getMessage());
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php:116]: > Should this message be moved to mftf.log instead like the rest of the errors and only display the total count of tests that failed to generate?
> Currently only "unable to create test" error is called out in the console specifically.
> 
> **Current:**
> In console:
> 
> > ERROR: Unable to create test AdminLoginSuccessfulTest
> > Mftf Test can not extend from itself: AdminLoginSuccessfulTest
> > ERROR: 8 Test failed to generate. See mftf.log for details.
> 
> **Expected:**
> In console:
> 
> > ERROR: 8 Test failed to generate. See mftf.log for details.
> 
> In mftf log:
> 
> > ERROR: Unable to create test AdminLoginSuccessfulTest
> > Mftf Test can not extend from itself: AdminLoginSuccessfulTest

Fixed

File: src/Magento/FunctionalTestingFramework/Util/GenerationErrorHandler.php [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                    }
+                }
+                $totalNotGenErrors = $totalErrors - $totalAnnotationErrors;
+                if ($totalNotGenErrors > 0) {
+                    print(
+                        PHP_EOL
+                        . 'ERROR: '
+                        . strval($totalNotGenErrors)
+                        . ' '
+                        . ucfirst($type)
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Util/GenerationErrorHandler.php:128]: Changed

File: src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php [hunk 12]
```diff
@@ -102,28 +102,37 @@ public function getObject($testName)
...
                 }
             }
         }
+
+        if ($errCount > 0
+            && MftfApplicationConfig::getConfig()->getPhase() == MftfApplicationConfig::GENERATION_PHASE) {
+            print(
+                "ERROR in TestObjectHandler::getAllObjects(): "
+                . strval($errCount)
+                . " Test(s) cannot to be extended. See mftf.log for details."
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php]: grammer fix:
`Test(s) cannot to be extended. See mftf.log for details. `

File: src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php [hunk 13]
```diff
@@ -169,12 +244,15 @@ private function initTestData($validateAnnotations = true)
...
      * @return TestObject
      * @throws TestFrameworkException
+     * @throws XmlException
      */
     private function extendTest($testObject)
     {
         if ($testObject->getParentName() !== null) {
             if ($testObject->getParentName() == $testObject->getName()) {
-                throw new TestFrameworkException("Mftf Test can not extend from itself: " . $testObject->getName());
+                throw new TestFrameworkException(
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php:282]: I think it's so too. It's not related to this PR.

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php [hunk 14]
```diff
@@ -88,31 +92,36 @@ public static function getInstance(): SuiteGenerator
...
-                LoggingUtil::getInstance()->getLogger(self::class)->notification(
-                    "Suite '" . $suiteName . "' contains no tests and won't be generated." . PHP_EOL,
-                    [],
-                    true
-                );
-                continue;
-            }
-            $firstElement = array_values($suiteContent)[0];
+            try {
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:102]: Good catch! Fixed.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php [hunk 15]
```diff
@@ -71,28 +81,80 @@ public function parseSuiteDataIntoObjects($parsedSuiteData)
...
-            };
+                // add all test if include tests is completely empty
+                if (empty($includeTests)) {
+                    $includeTests = TestObjectHandler::getInstance()->getAllObjects();
+                }
 
-            // add all test if include tests is completely empty
-            if (empty($includeTests)) {
-                $includeTests = TestObjectHandler::getInstance()->getAllObjects();
+                if (!empty($includeMessage)) {
```
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:127]: Fixed

File: src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php (hunk 12 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Test/Handlers/TestObjectHandler.php]: Fixed

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php [hunk 16]
```diff
@@ -140,35 +149,76 @@ public function generateSuite($suiteName)
...
-        $this->generateRelevantGroupTests($suiteName, $relevantTests);
-        $groupNamespace = $this->generateGroupFile($suiteName, $relevantTests, $originalSuiteName);
+            try {
+                $this->generateRelevantGroupTests($suiteName, $relevantTests);
+            } catch (FastFailException $e) {
+                throw $e;
+            } catch (\Exception $e) {
+                $exceptionCollector->addError(
+                    self::class,
+                    "Failed to generate tests for suite \"{$suiteName}\""
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: I introduced a bad entity reference in a test that is a part of WYSIWYGDisabledSuite, but I don't see this error message on the console. mftf.log looks good to me, and the test did not generate, which looks good to.

```
[2020-09-24 15:47:23] Magento\FunctionalTestingFramework\Util\TestGenerator.ERROR: Failed to generate NewProductsListWidgetBundleProductTest [] []
[2020-09-24 15:47:23] Magento\FunctionalTestingFramework\Test\Handlers\TestObjectHandler.ERROR: Removed invalid test object NewProductsListWidgetBundleProductTest [] []
```

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php [hunk 17]
```diff
@@ -71,28 +81,80 @@ public function parseSuiteDataIntoObjects($parsedSuiteData)
...
+                // it's ok if failed in exclude
+                $exclude = $this->extractTestObjectsFromSuiteRef($groupTestsToExclude);
+                $excludeTests = $exclude['objects'] ?? [];
+
+                // parse any object hooks
+                $suiteHooks = $this->parseObjectHooks($parsedSuite);
+
+                // log error if suite is empty
+                if ($this->isSuiteEmpty($suiteHooks, $includeTests, $excludeTests)) {
+                    LoggingUtil::getInstance()->getLogger(self::class)->error(
```
Comment (soumyau) [inline src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:109]: Should unable to parse suite be moved to mftf.log as well, for consistent error logging?
 
Current:
```
NOTICE: 202 Section name violations detected. See mftf.log for details.[]
NOTICE: 617 section element name violations detected. See mftf.log for details.[]
NOTICE: 301 data entity name violations detected. See mftf.log for details.[]
NOTICE: 5 data entity key violations detected. See mftf.log for details.[]
NOTICE: 10 action group name violations detected. See mftf.log for details.[]
ERROR: Unable to parse suite WYSIWYGDisabledSuite
NOTICE: 38 Page name violations detected. See mftf.log for details.[]

ERROR: 1 Suite(s) failed to generate. See mftf.log for details.
ERROR: 7 Test(s) failed to generate. See mftf.log for details.
Generate Tests Command Run (with errors)
```

Expected:
`ERROR: Unable to parse suite WYSIWYGDisabledSuite` only appears in mftf.log

**Comment by jilu1** [comment]
UR Build:
https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/36062/

Two test failures seem non related.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 17 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:109]: We are logging some errors to console before and now, like deprecation message, story/title uniqueness, some extends error, etc. I feel display these errors are useful as long as they are not all over the place. The log, on the other hand, is quite messy and hard to find.

I guess the question is:
Not only this suite message, there are other error messages. Do we want to remove all the error display?

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 16 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: what is the expected behavior if we try to generate a non-existant suite
I ran this:
`bin/mftf generate:suite NonExistant suite`
What I see is all tests are generated except for the ones with errors. That shouldn't be the case, should it? On `develop` the command throws an error, and no tests get generated.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 17 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:109]: I get your point. I'm okay with this too. Looks like there are other error messages as well that get displayed. There many be some inconsistency on which errors are displayed on console and which are not. It may be good to list them down to see which ones are useful to be printed in console and which are not.

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 16 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: I cannot reproduce this one it seems. I got 
> ERROR: 2 Test(s) generated with annotation errors. See mftf.log for details.
> ERROR: 1 Suite(s) failed to generate. See mftf.log for details.
> No Suite Generated

Can you specify the steps?

Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: > I introduced a bad entity reference in a test that is a part of WYSIWYGDisabledSuite, but I don't see this error message on the console. mftf.log looks good to me, and the test did not generate, which looks good to.
> 
> ```
> [2020-09-24 15:47:23] Magento\FunctionalTestingFramework\Util\TestGenerator.ERROR: Failed to generate NewProductsListWidgetBundleProductTest [] []
> [2020-09-24 15:47:23] Magento\FunctionalTestingFramework\Test\Handlers\TestObjectHandler.ERROR: Removed invalid test object NewProductsListWidgetBundleProductTest [] []
> ```

This is expected and result of the change. When a bad test is encountered, it's removed from the test objects and it's not discovered by the suite.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 17 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:109]: I think it might be good to remove all error displays. This way we can force users to look at the log for errors.

File: src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php (hunk 16 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: yes, but shouldn't it increment the number of tests(s) failed to generate in the console message?

Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: > I cannot reproduce this one it seems. I got
> 
> > ERROR: 2 Test(s) generated with annotation errors. See mftf.log for details.
> > ERROR: 1 Suite(s) failed to generate. See mftf.log for details.
> > No Suite Generated
> 
> Can you specify the steps? And there is no test or suite generated.

I am unable to as well. Maybe some tests that were left behind in a previous run

Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: > yes, but shouldn't it increment the number of tests(s) failed to generate in the console message?
Only part left to verify is the above, from these comments

Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/SuiteGenerator.php:195]: The bad test will be removed and counted for a failure, but the suite will not know there is this failing test. This is ok.

File: src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php (hunk 17 above)
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Suite/Util/SuiteObjectExtractor.php:109]: > I think it might be good to remove all error displays. This way we can force users to look at the log for errors.

Implement it as default. Error will print in verbose mode.

File: docs/commands/mftf.md [hunk 18]
```diff
@@ -89,6 +89,15 @@ vendor/bin/mftf run:failed
 This command cleans up the previously generated tests; generates and runs the tests listed in `dev/tests/acceptance/tests/_output/failed`.
 For more details about `failed`, refer to [Reporting][].
 
+## Error tolerance during generation
+
+Starting from version 3.2.0, MFTF will not fail right away when encounter generation errors.
```
Comment (dobooth) [inline docs/commands/mftf.md:94]: 1. 'encountering'

2. Instead, MFTF will generate as many tests and suites as it can, log errors to `mftf.log`, and exit with a non-zero generation status.

3. For example, schema validation errors, parser errors, and WebApi authentication errors will cause `hard` failures, with no tests or suites being generated.

4. Error tolerance in generation is meant to help local test development and testing and are expected to be run locally. All generation errors must be fixed in order to use other framework functionality, pass static checks, and to deliver MFTF tests.

**Review by dobooth — CHANGES_REQUESTED** [review]

**Review by dobooth — APPROVED** [review]

**Comment by tomreece** [comment]
@magento import PR to magento-commerce

**Comment by magento-engcom-team** [comment]
@okolesnyk the pull request successfully imported.
